
# Scripts for Unannotated Product Tests

SmallRNA sequencing reads were mapped using **miRWoods/createBam&#46;pl** and identification of unnannotated microRNAs was performed using **miRWoods/miRPreprocess&#46;pl**.  Both of these scripts are found in the [miRWoods repository](https://github.com/JimBell/miRWoods).

Analysis of Unannotated products was performed by running scripts in the following order:

**getOneProductMiRFasta\_forSpecies&#46;pl** takes the deepMirCut dataSet file (mir\_dataset.txt), the mirbase hairpin precursor file (hairpin.fa), the mirbase gff file for human mirs (hsa.gff3), and the species prefix for human (hsa) as input.  It outputs a fasta file with precursers for mirs with only one annotated product in hsa_OneProduct.fa
```sh
$ perl Scripts/getOneProductMiRFasta_forSpecies.pl mir_dataset.txt hairpin.fa hsa.gff3 hsa
```

**compareMaxNeedleIdentity&#46;pl** takes a fasta file of human mirs with only one annotated product (hsa\_OneProduct.fa), a fasta file of mirs in the training set without the buffer regions (trainSet\_wSimilar\_precursors.fa), and a output prefix (hsa\_vs\_trainSetWSimilar).  It finds the maximum identity for each mir in hsa\_OneProduct.fa comapred to all mirs in trainSet\_wSimilar\_precursors.fa, and outputs a file listing the maximum identities in hsa_vs_trainSetWSimilar.txt
```sh
$ perl Scripts/compareMaxNeedleIdentity.pl hsa_OneProduct.fa trainSet_wSimilar_precursors.fa hsa_vs_trainSetWSimilar
```

**createOneProductTestSet&#46;pl** takes the deepMirCut dataSet file (mir\_dataset.txt), and the file containing the maximum identities in hsa\_vs\_trainSetWSimilar.txt and outputs a test set file of microRNAs with identities less than 80 percent (test\_oneProd.txt)
```sh
$ perl Scripts/createOneProductTestSet.pl mir_dataset.txt hsa_vs_trainSetWSimilar.txt
```

**addFoldsToSet&#46;pl** takes the test set (test\_oneProd.txt) as input, and outputs test set file with dot-bracket folds and a bpRNA context array (test\_oneProd\_wFolds.txt)
```sh
$ perl Scripts/addFoldsToSet.pl test_oneProd.txt
```

**prepareNewCutsiteSets&#46;pl** takes the test set with dot-bracket frolds and bpRNA context array (test\_oneProd\_wFolds.txt), the hairpin file output by miRPreprocess.pl (readRegions\_hairpins.txt), the products file output by miRPreprocess.pl (readRegions\_products.txt), and a number indicating the minimum number of reads (in this case 5).  It adds cutsites for the unannotated products using the files generated by miRPreprocess, and then outputs two test sets based on which side the unannotated product was on (oneProd\_test\_3pAnnot\_5pUnannot.txt, oneProd\_test\_5pAnnot_3pUnannot.txt)
```sh
$ perl Scripts/prepareNewCutsiteSets.pl test_oneProd_wFolds.txt readRegions_hairpins.txt readRegions_products.txt 5
```

Test sets were evaluated with **deepMirCut/testModel&#46;py** using the ensemble with the -d flag set to output decision values.  (You may need to change the path depending on where you have deepMirCut/testModel.py installed.)
```sh
$ python3 ~/deepMirCut/testModel.py oneProd_test_3pAnnot_5pUnannot.txt -L seqBPRNA_ensemble_list.txt -o 3pAnnot_5pUnannot -d --input_setting 2
$ python3 ~/deepMirCut/testModel.py oneProd_test_5pAnnot_3pUnannot.txt -L seqBPRNA_ensemble_list.txt -o 5pAnnot_3pUnannot -d --input_setting 2
```

**graph\_deltas\_histogram\_unannot&#46;py** takes the test sets (oneProd\_test\_3pAnnot\_5pUnannot.txt, oneProd\_test\_5pAnnot\_3pUnannot.txt) and the decision values returned by deepMirCut (3pAnnot\_5pUnannot\_classification\_DVs.txt, 5pAnnot\_3pUnannot\_classification\_DVs.txt) and an output prefix (unannotated) as input.  It produces a graph (unannotated_hist.svg) as output.
```sh
$ python3 Scripts/graph_deltas_histogram_unannot.py oneProd_test_3pAnnot_5pUnannot.txt oneProd_test_5pAnnot_3pUnannot.txt 3pAnnot_5pUnannot_classification_DVs.txt 5pAnnot_3pUnannot_classification_DVs.txt unannotated
```

**getOneProductTestSetReads&#46;pl** takes the deepMirCut dataset (mir\_dataset.txt), the test file (oneProd\_test\_3pAnnot\_5pUnannot.txt or oneProd\_test\_5pAnnot\_3pUnannot.txt), a fasta of the human genome (hg38.fa), and a bam list containing the locations of bam files with reads mapped to the genome (bamList.txt).  It outputs a file with a list of read counts 
```sh
$ perl Scripts/getOneProductTestSetReads.pl mir_dataset.txt oneProd_test_3pAnnot_5pUnannot.txt hg38.fa bamList.txt
$ perl Scripts/getOneProductTestSetReads.pl mir_dataset.txt oneProd_test_5pAnnot_3pUnannot.txt hg38.fa bamList.txt
```

**graphOneProductReads&#46;py** takes a file with a list of read counts, the test set file, a file with decision values output by deepMirCut, and a side (5p or 3p).  It outputs a readcount graph as output.
```sh
$ python3 Scripts/graphOneProductReads.py oneProd_test_3pAnnot_5pUnannot.txt.counts oneProd_test_3pAnnot_5pUnannot.txt 3pAnnot_5pUnannot_classification_DVs.txt 5p
$ python3 Scripts/graphOneProductReads.py oneProd_test_5pAnnot_3pUnannot.txt.counts oneProd_test_5pAnnot_3pUnannot.txt 5pAnnot_3pUnannot_classification_DVs.txt 3p
```


